google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkPanel.java

```diff
@@ -70,46 +71,85 @@ public CloudSdkPanel() {
     if (StringUtil.isEmpty(path)) {
-      warningIcon.setVisible(true);
-      warningMessage.setVisible(true);
-      warningMessage.setText(
-          createErrorMessageWithLink(
-              GctBundle.message("appengine.cloudsdk.location.missing.message")));
-
+      String warningText = appendCloudSdkDownloadMessage(
+              GctBundle.message("appengine.cloudsdk.location.missing.message"));
+      showWarning(warningText, false /* setSdkDirectoryErrorState */);
       return;
     }
 
-    try {
-      new CloudSdk.Builder()
-          .sdkPath(Paths.get(path))
-          .build()
-          .validateCloudSdk();
-
-      cloudSdkDirectoryField.getTextField().setForeground(JBColor.black);
-      warningIcon.setVisible(false);
-      warningMessage.setVisible(false);
-    } catch (AppEngineException aee) {
-      cloudSdkDirectoryField.getTextField().setForeground(JBColor.red);
-      warningIcon.setVisible(true);
-      warningMessage.setVisible(true);
-      warningMessage.setText(
-          createErrorMessageWithLink(
-              GctBundle.message("appengine.cloudsdk.location.invalid.message")));
+    Set<CloudSdkValidationResult> validationResults =
+        CloudSdkService.getInstance().validateCloudSdk(Paths.get(path));
+
+    // if there are any validation errors, show the first one to the user
+    if (validationResults.size() > 0) {
+      CloudSdkValidationResult validationResult = validationResults.iterator().next();
+      String message;
+      switch (validationResult) {
+        case CLOUD_SDK_NOT_FOUND:
+          message = appendCloudSdkDownloadMessage(validationResult.getMessage());
+          break;
+        case CLOUD_SDK_VERSION_NOT_SUPPORTED:
+          message = validationResult.getMessage();
+          break;
+        default:
+          message = validationResult.getMessage();
+      }
+      showWarning(message, validationResult.isError());
+    } else {
+      hideWarning();
     }
   }
 
   @VisibleForTesting
+  protected void showWarning(final String message, final boolean setSdkDirectoryErrorState) {
+    invokePanelValidationUpdate(new Runnable() {
+      @Override
+      public void run() {
+        warningIcon.setVisible(true);
+        warningMessage.setVisible(true);
+        warningMessage.setText(message);
+
+        if (setSdkDirectoryErrorState) {
+          cloudSdkDirectoryField.getTextField().setForeground(JBColor.red);
+        }
+      }
+    });
+  }
+
+  @VisibleForTesting
+  protected void hideWarning() {
+    invokePanelValidationUpdate(new Runnable() {
+      @Override
+      public void run() {
+        cloudSdkDirectoryField.getTextField().setForeground(JBColor.black);
+        warningIcon.setVisible(false);
+        warningIcon.setVisible(false);
+        warningMessage.setVisible(false);
+      }
+    });
+  }
+
+  private void invokePanelValidationUpdate(Runnable runnable) {
+    ApplicationManager.getApplication().invokeLater(runnable,
```

Comment:
No, because showWarning and hideWarning are called within `checkSdk` which is always called by a background thread. See `checkSdkInBackground`.